### PR TITLE
Support wsl.

### LINKED ISSUE #8: if-watch does not work on windows/WSL2
On the "linux" provided by WSL2 windows, if-watch does not work. It bails out with an OS Error 22 - illegal arguments. Presumably because the API used by if-watch is not supported on the WSL2 flavour of "linux".

It would be useful to get this to work, by using the poll based fallback in all cases where the event based approach is not supported by the operating system.

E.g. this causes this issue in ipfs-embed:
https://github.com/ipfs-rust/ipfs-embed/issues/63

### PR DESCRIPTION
**Pull request by dvc94ch**
Closes #8

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -36,25 +32,40 @@ pub enum IfEvent {
...
+    /// Polling fallback.
+    Fallback(fallback::IfWatcher),
+}
 
 impl IfWatcher {
     /// Create a watcher
     pub async fn new() -> Result<Self> {
-        Ok(Self(platform_impl::IfWatcher::new().await?))
+        if std::env::var_os("WSL_DISTRO_NAME").is_none() {
+            Ok(Self::Platform(platform_impl::IfWatcher::new().await?))
```
Comment (2beaucoup) [inline src/lib.rs:46]: not sure where exactly this failed before, but does it make sense to try the fallback if this errored?

File: src/lib.rs [hunk 2]
```diff
@@ -36,25 +32,40 @@ pub enum IfEvent {
...
+    Platform(platform_impl::IfWatcher),
+    /// Polling fallback.
+    Fallback(fallback::IfWatcher),
+}
 
 impl IfWatcher {
     /// Create a watcher
     pub async fn new() -> Result<Self> {
-        Ok(Self(platform_impl::IfWatcher::new().await?))
+        if std::env::var_os("WSL_DISTRO_NAME").is_none() {
```
Comment (rklaehn) [inline src/lib.rs:45]: I got a feeling that there will be more weird linux situations where the event based stuff does not work. So I am not really opposed to just trying and using the fallback on error.

But since so far we know of only this case, I am also fine with checking for it.

Let's revisit it once this becomes a longer list.

File: src/fallback.rs [hunk 3]
```diff
@@ -51,7 +51,7 @@ impl IfWatcher {
         Ok(())
     }
 
-    pub fn iter(&self) -> impl Iterator<Item = &IpNet> {
+    pub fn iter(&self) -> Iter<'_, IpNet> {
```
Comment (rklaehn) [inline src/fallback.rs:54]: What's the reason for this change?

**Review by rklaehn — APPROVED** [review]
So the check is done at runtime? Works for me.

**Comment by rklaehn** [comment]
FYI:

not sure what the tests do, but they are green on the wsl branch, and red on master, on WSL2

File: src/fallback.rs (hunk 3 above)
Comment (dvc94ch) [reply src/fallback.rs:54]: because the fallback and platform_impl need to return the same iterator

Comment (dvc94ch) [reply src/fallback.rs:54]: rust can't assume that it is the same iterator when using impl even if it is

File: src/lib.rs (hunk 1 above)
Comment (dvc94ch) [reply src/lib.rs:46]: this might lead to missing bug reports, so I'd prefer not to do it
